File: dart/knapsack/lib/knapsack.dart
import 'dart:math';

typedef Item = ({int weight, int value});

class Knapsack {
  final int maxWeight;
  Knapsack({required this.maxWeight});

  // Using the algorithm at
  // https://en.wikipedia.org/wiki/Knapsack_problem#0-1_knapsack_problem

  int maxValue(List<Item> items) {
    final m =
        List.generate(items.length + 1, (_) => List.filled(maxWeight + 1, 0));

    var i = 0;
    for (final item in items) {
      i += 1;
      for (var w = 0; w <= maxWeight; w++)
        m[i][w] = item.weight > w
            ? m[i - 1][w]
            : max(m[i - 1][w], item.value + m[i - 1][w - item.weight]);
    }
    return m[items.length][maxWeight];
  }
}

File: ruby/knapsack/knapsack.rb
# rubocop:disable Style/FrozenStringLiteralComment

## Using the [0-1 knapsack problem][1] as specified on wikipedia
## [1]: https://en.wikipedia.org/wiki/Knapsack_problem#0-1_knapsack_problem
##
class Knapsack
  def initialize(max_weight)
    @max_weight = max_weight
  end

  def max_value(items)
    values = Array.new(items.count + 1) { Array.new(@max_weight + 1, 0) }
    items.each_with_index do |item, i|
      1.upto(@max_weight) do |wt|
        val = [values[i][wt]]
        val << (values[i][wt - item.weight] + item.value) if item.weight <= wt
        values[i + 1][wt] = val.max
      end
    end
    values[items.count][@max_weight]
  end
end

# rubocop:enable Style/FrozenStringLiteralComment
